code/game/objects/items/stacks/rods.dm

```diff
@@ -60,30 +67,53 @@ GLOBAL_LIST_INIT(rod_recipes, list ( \
 	else
 		icon_state = "rods"
 
-/obj/item/stack/rods/attackby(obj/item/W, mob/user, params)
-	if(W.tool_behaviour == TOOL_WELDER)
+/obj/item/stack/rods/attackby(obj/item/attackby_item, mob/user, params)
+	if(attackby_item.tool_behaviour == TOOL_WELDER)
 		if(get_amount() < 2)
-			to_chat(user, span_warning("You need at least two rods to do this!"))
+			balloon_alert(user, "not enough rods!")
 			return
-
-		if(W.use_tool(src, user, 0, volume=40))
+		if(attackby_item.use_tool(src, user, 0, volume = 40))
 			var/obj/item/stack/sheet/iron/new_item = new(usr.loc)
-			user.visible_message(span_notice("[user.name] shaped [src] into iron sheets with [W]."), \
-				span_notice("You shape [src] into iron sheets with [W]."), \
-				span_hear("You hear welding."))
-			var/obj/item/stack/rods/R = src
+			balloon_alert(user, "crafting sheets...")
+			user.visible_message(
+				span_notice("[user.name] shaped [src] into iron sheets with [attackby_item]."),
+				blind_message = span_hear("You hear welding."),
+				vision_distance = COMBAT_MESSAGE_RANGE,
+				ignored_mobs = user
+			)
+			var/obj/item/stack/rods/welded_rod = src
 			src = null
-			var/replace = (user.get_inactive_held_item()==R)
-			R.use(2)
-			if (!R && replace)
-				user.put_in_hands(new_item)
+			welded_rod.use(2)
+			user.put_in_inactive_hand(new_item)
 	else
 		return ..()
 
+/obj/item/stack/rods/attackby_secondary(obj/item/attackby_item, mob/user, params)
```

Comment:
``welder_act_secondary()`` please

Reply:
I had no idea that even existed but it made things a lot easier for me, thonks!